subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/processing/AnnotationProcessorDetector.java

```diff
@@ -0,0 +1,128 @@
+/*
+ * Copyright 2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.tasks.compile.processing;
+
+import com.google.common.base.Charsets;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Lists;
+import com.google.common.io.CharStreams;
+import com.google.common.io.Files;
+import org.apache.tools.zip.ZipEntry;
+import org.apache.tools.zip.ZipFile;
+import org.gradle.api.UncheckedIOException;
+import org.gradle.cache.internal.FileContentCache;
+import org.gradle.cache.internal.FileContentCacheFactory;
+import org.gradle.internal.FileUtils;
+import org.gradle.internal.file.FileType;
+import org.gradle.internal.serialize.ListSerializer;
+import org.gradle.util.DeprecationLogger;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * Inspects a classpath to find annotation processors contained in it.
+ */
+public class AnnotationProcessorDetector {
+    private final FileContentCache<List<AnnotationProcessorDeclaration>> cache;
+
+    public AnnotationProcessorDetector(FileContentCacheFactory cacheFactory) {
+        cache = cacheFactory.newCache("annotation-processors", 20000, new ProcessorServiceLocator(), new ListSerializer<AnnotationProcessorDeclaration>(new AnnotationProcessorDeclarationSerializer()));
+    }
+
+    public List<AnnotationProcessorDeclaration> detectProcessors(Iterable<File> processorPath) {
+        List<AnnotationProcessorDeclaration> processors = Lists.newArrayList();
```

Comment:
This is nitpicking but we could avoid creating a list if there's no processor, which should be pretty common.

Reply:
I considered that, but this is called so infrequently that it's unlikely to make a difference. Plus it makes the code more awkward to read.